### ISSUE TITLE
Cannot find some modules when use npm@5.2.0 install standard

### ISSUE THREAD

**Issue by LitoMore** [npm/npm#17717]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

Some modules is missing when I use `npm i -D standard`.
You can find `eslint-plugin-react` module in `./node_modules/standard/node_modules/`.

```
standard: Unexpected linter output:

Error: Failed to load plugin react: Cannot find module 'eslint-plugin-react'
Referenced from: /Users/litomore/Git/npm-5.2.0-issue/node_modules/standard/eslintrc.json
    at Function.Module._resolveFilename (module.js:485:15)
    at Function.resolve (internal/module.js:18:19)
    at Object.load (/Users/litomore/Git/npm-5.2.0-issue/node_modules/eslint/lib/config/plugins.js:134:29)
    at Array.forEach (native)
    at Object.loadAll (/Users/litomore/Git/npm-5.2.0-issue/node_modules/eslint/lib/config/plugins.js:162:21)
    at load (/Users/litomore/Git/npm-5.2.0-issue/node_modules/eslint/lib/config/config-file.js:541:21)
    at configExtends.reduceRight (/Users/litomore/Git/npm-5.2.0-issue/node_modules/eslint/lib/config/config-file.js:424:36)
    at Array.reduceRight (native)
    at applyExtends (/Users/litomore/Git/npm-5.2.0-issue/node_modules/eslint/lib/config/config-file.js:408:28)
    at Object.load (/Users/litomore/Git/npm-5.2.0-issue/node_modules/eslint/lib/config/config-file.js:566:22)

If you think this is a bug in `standard`, open an issue: https://github.com/feross/standard/issues
```

#### How can the CLI team reproduce the problem?

1. `mkdir npm-5.2.0-issue && cd npm-5.2.0-issue`
2. `npm init -y && npm i -D standard`
3. `./node_modules/.bin/standard`

### supporting information:

 - `npm -v` prints: 5.2.0
 - `node -v` prints: v8.1.2
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: macOS
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [x] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by yvele** [comment]
Looks like it's related to new 5.2.0 version, specially this PR https://github.com/npm/npm/pull/17678#issuecomment-314397354

> 06c154fd6 #17678 Stop flattening modules that have peer dependencies. We're making this change to support scenarios where the module requiring a peer dependency is flattened but the peer dependency itself is not, due to conflicts. In those cases the module requiring the peer dep can't be flattened past the location its peer dep was placed in. This initial fix is naive, never flattening peer deps, and we can look into doing something more sophisticated later on. (@iarna)


-------

In my use case I have the following problem:


```
Oops! Something went wrong! :(

ESLint couldn't find the plugin "eslint-plugin-babel". This can happen for a couple different reasons:

1. If ESLint is installed globally, then make sure eslint-plugin-babel is also installed globally. A globally-installed ESLint cannot find a locally-installed plugin.

2. If ESLint is installed locally, then it's likely that the plugin isn't installed correctly. Try reinstalling by running the following:

    npm i eslint-plugin-babel@latest --save-dev

If you still can't figure out the problem, please stop by https://gitter.im/eslint/eslint to chat with the team.
```

> 
> This totally breaks my whole organisation 😭 
> 
> I have some packages that also acts like "_dependency presets_".. When you get one, you also get his **flattened** dependencies, and I'm depending on them.
> 
> In my use case, the problem is that many packages like [eslint-plugin-babel](https://github.com/babel/eslint-plugin-babel/blob/master/package.json) have peer dependencies and are not flattened anymore... I cannot use my "_dependency presets_" anymore, and instead should explicitly list ALL dependencies
> 
> Should I open an issue? Is my use of "_dependency presets_" not legit?

**Comment by fmal** [comment]
I have the same issue in 5.2.0. Reverting to previous NPM version and doing fresh install generates different tree with flattened eslint plugins.

**Comment by swernerx** [comment]
Same issue for be... reporting babel-loader not found.

**Comment by iowillhoit** [comment]
Same issue. We have a lot of core packages that "inherit" dependencies from each other. Some of those have peer dependencies, so they no longer get flattened to the root `node_modules` directory. 

This seems like a large change for a minor version. I feel like this should have been an option to start with, something like: `--dont-flatten-peers`.

**Comment by iarna** [comment]
I've reverted the commit that caused this as doing this right is rather more complicated and the bug it was fixing was way more of a corner case than this.

The fix that will actually work for people is for a peer dep to insist on never being installed deeper than the the thing it relies on.  (Which is... kind of the opposite of the patch.) At the moment this is tricky because the thing the thing with the peer dep relies on may not yet have been added to the tree at the time that we're placing the thing with the peer dep, which means we can't make decisions based on where that thing will be placed. Er, if that makes sense.

Anyway, I'm declaring that peer deps are still a ball of hornets and I'm gonna avoid touching them again till we have a spec and spec test for all of the complicated corner cases.
